app/services/fact_importer.rb

```diff
@@ -66,11 +68,17 @@ def add_new_facts
     facts_to_create = facts.keys - db_facts.keys
     # if the host does not exists yet, we don't have an host_id to use the fact_values table.
     if facts_to_create.present?
-      method          = host.new_record? ? :build : :create!
-      fact_names      = fact_name_class.group(:name).maximum(:id)
+      method = host.new_record? ? :build : :create!
+      # :type is needed because custom facts usually inherits from FactName so they would be included in the list
+      fact_names = fact_name_class.where(:type => fact_name_class).group(:name).maximum(:id)
       facts_to_create.each do |name|
-        host.fact_values.send(method, :value => facts[name],
-                              :fact_name_id  => fact_names[name] || fact_name_class.create!(:name => name).id)
+        begin
+          host.fact_values.send(method, :value => facts[name],
+                                :fact_name_id  => fact_names[name] || fact_name_class.create!(:name => name).id)
```

Comment:
Do I understand it correctly, that this validator got triggered https://github.com/theforeman/foreman/blob/57e9d8a360625afb6dfe292655cc03cd65719073/app/models/fact_value.rb#L37 when we don't explicitly filter on the specific line here https://github.com/theforeman/foreman/pull/3465/files#diff-671e5385f8ec83d391c87d7c292604edR73